src/main/java/com/avispl/symphony/dal/communicator/middleatlantic/MiddleAtlanticPowerUnitCommunicator.java

```diff
@@ -39,56 +42,63 @@ public MiddleAtlanticPowerUnitCommunicator() {
         // it can be changed back by configuration
         setTrustAllCertificates(true);
         localStatistics = new ExtendedStatistics();
-        localStatistics.setControl(new HashMap<>());
-        localStatistics.setStatistics(new HashMap<>());
     }
 
     @Override
     protected void authenticate() throws Exception {
         // nothing to do here, authentication is done in individual requests
     }
 
+    /**
+     * Getting statistics of the device.
+     * Need to make sure that none of control operations are running when trying to fetch the statistics
+     */
     @Override
     public List<Statistics> getMultipleStatistics() throws Exception {
+        ExtendedStatistics extendedStatistics = new ExtendedStatistics();
         // This is to make sure if the statistics is being fetched before/after any set of control operations
-        if(!controlsRunning) {
-            try {
-                readWriteLock.writeLock().lock();
-                ExtendedStatistics extendedStatistics = new ExtendedStatistics();
-                Map<String, String> statistics = Collections.synchronizedMap(new LinkedHashMap<>());
-                Map<String, String> control = new ConcurrentHashMap<>();
-
-                ExecutorService executor = Executors.newCachedThreadPool();
-                runAsync(() -> fillInStatistics(statistics, "Inlet Active Power", "/inlet/0/activePower", GET_READING), executor)
-                        .thenRunAsync(() -> fillInStatistics(statistics, "Inlet RMS Current", "/inlet/0/current", GET_READING), executor)
-                        .thenRunAsync(() -> fillInStatistics(statistics, "Inlet Line Frequency", "/inlet/0/lineFrequency", GET_READING), executor)
-                        .thenRunAsync(() -> fillInStatistics(statistics, "Inlet RMS Voltage", "/inlet/0/voltage", GET_READING), executor)
-                        .thenApplyAsync(ignore -> getOutletsCount(), executor)
-                        .thenAcceptAsync(count -> IntStream.range(0, count)
-                                .parallel()
-                                .mapToObj(num -> runAsync(() -> fillInOutletState(statistics, control, num), executor)
-                                        .thenRunAsync(() -> fillInStatistics(statistics, getOutletRMSName(num),
-                                                "/outlet/" + num + "/current", GET_READING), executor)
-                                )
-                                .forEach(CompletableFuture::join), executor)
-                        .get(30, TimeUnit.SECONDS);
-
-                executor.shutdownNow();
-
-                extendedStatistics.setStatistics(statistics);
-                extendedStatistics.setControl(control);
-                localStatistics = extendedStatistics;
-            } finally {
-                readWriteLock.writeLock().unlock();
-            }
+        if(controlsRunning){
+            extendedStatistics.setStatistics(localStatistics.getStatistics());
```

Comment:
That does not solve issue with mutable maps. This will return the same mutable maps (from localStatistics) to a caller, just now as a property of extendedStatistics object. If caller happens to modify the map, that modifications will be reflected in localStatistics. If you want to resolve it, you need to create copy of maps here (instead of using ones from localStatistics).